### PR TITLE
Construct workers out of Components

### PR DESCRIPTION
**Pull request by abyrd**
The goal of this PR is to use our framework-free dependency injection approach, assembling the workers out of Components the same way we do the backend.

Besides simple uniformity, this also has other practical advantages. It will allow us to use the TaskScheduler component to facilitate the regular polling described in #596. It will also allow us to switch to a uniform storage abstraction, using standard FileStorage components, so we can remove deployment environment-specific dependencies from R5.

As of 2021-03-12 this is under active development. Expecting to convert from draft this week.

### CODE REVIEW

**Comment by abyrd** [comment]
This should be made compatible with the proposed changes under discussion in #694

**Comment by abyrd** [comment]
I have done two new rounds of changes on this PR.

First, I refactored the Config classes to unify the approach used for backend and worker configuration. The commits that complete this stage of the refactoring are e26529d...7b562f1. Rather than just the commit diffs, I'd recommend looking at the resulting files to clearly see the end result. For example, see BackendConfig https://github.com/conveyal/r5/blob/7b562f18d63ee8212e8b826e8515dddd8f419b59/src/main/java/com/conveyal/analysis/BackendConfig.java
and WorkerComponents https://github.com/conveyal/r5/blob/7b562f18d63ee8212e8b826e8515dddd8f419b59/src/main/java/com/conveyal/analysis/components/WorkerComponents.java

Second, I applied my perspective on #702, which I described at a high level in a recent meeting. I see the underlying problem as passing around config information (bucket names) for things that do not need to be free-text options, and can be systematically derived from compile-time constants (the category of files handled by each component) and runtime constants (the deployment environment).

Basically, the filesystem or bucket directories are treated as an encapsulated detail of FileStorage implementations. Rather than passing arbitrary bucket names as Strings into various components and sub-components, we just have each component specify what category of files it saves via compile-time constants. The FileStorage implementation maps file types to appropriate storage concepts. So for local storage the file categories correspond to subdirectories of the cache directory; for cloud storage the file categories become suffixes appended to a common bucket prefix. 

This can be applied directly to our existing deployments without a storage migration, because common prefixes are already in use and implied by our deployment environment. Looking at our current staging and production config properties, we see that the bucket parameters are identical between the two and distinguished only by a prefix. 

If and when we do decide to migrate to single top-level buckets (e.g. conveyal-production/results instead of scenario-editor-prod-results) we're basically just changing a dash into a slash and changing the prefix.

These changes are fairly complete, but have only been tested locally and via the Cypress workflow (but not yet in a cloud deployment). Anyway they still need some work to fully differentiate the cloud implementations of various components.
 
Again I'd advise looking at the resulting Components and Config files, not just the diffs, to appreciate how clean the result is. For example, https://github.com/conveyal/r5/blob/9f88adf74d170b09cead03f3af3f102714a0bd27/src/main/java/com/conveyal/analysis/BackendConfig.java and https://github.com/conveyal/r5/blob/9f88adf74d170b09cead03f3af3f102714a0bd27/src/main/java/com/conveyal/analysis/components/LocalBackendComponents.java become simpler with less small interfaces, and many components requiring no Config at all. The grid and CSV result writer classes no longer pass around any outpuBucket parameters.

I expect to do a few more rounds of simplifications and refactoring on this. If you want to try this locally on existing data, remove the prefixes from all the directory names in your cache directory.

File: src/main/java/com/conveyal/r5/analyst/cluster/WorkerMain.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+/**
+ * Old R5 JARs before December 2020 had main-class set to R5Main, which used the first command line parameter to
+ * select between starting an analysis worker and a point-to-point routing server. We no longer use the latter option,
+ * but we want to allow the backend to start all worker versions in exactly the same way, so we ignore the command line
+ * parameters. If we ever want to allow other roles, we can check that this parameter equals 'worker'.
+ */
+public class WorkerMain {
+
+    public static void main (String[] args) {
+        System.out.println("=== Conveyal R5 ===");
```
Comment (ansoncfit) [inline src/main/java/com/conveyal/r5/analyst/cluster/WorkerMain.java:16]: _important_: loss of ASCII art?

**Review by ansoncfit — COMMENTED** [review]
Nice, it is indeed much cleaner not to pass around `outputBucket` and other parameters excessively (e.g. in `MultiOriginAssembler.java`)

In many of the updated files, `import com.conveyal.file.FileCategory;` appears to be unused. Is that intentional?

File: src/main/java/com/conveyal/r5/analyst/cluster/WorkerMain.java (hunk 1 above)
Comment (abyrd) [reply src/main/java/com/conveyal/r5/analyst/cluster/WorkerMain.java]: I do like ASCII art but I think I often saw it garbled in logs.

**Review by ansoncfit — COMMENTED** [review]
Testing regional analysis and Taui site creation with `v6.2-testing` on staging has gone well. I added two comments to consider, after which this should be ready for merge.

File: src/main/java/com/conveyal/r5/analyst/cluster/EC2Info.java [hunk 2]
```diff
@@ -1,18 +1,10 @@
 package com.conveyal.r5.analyst.cluster;
```
[the remark this answers is not in the document]
Comment (ansoncfit) [reply src/main/java/com/conveyal/r5/analyst/cluster/EC2Info.java:1]: Ah, now I see the use of `privateIp` in `WorkerStatus`.  Maybe the other fields are reported to the admin dashboard, but I'm not sure they're useful there.

File: src/main/java/com/conveyal/analysis/BackendConfig.java [hunk 3]
```diff
@@ -4,173 +4,95 @@
...
 
-    public static final String CONVEYAL_PROPERTY_PREFIX = "conveyal-";
+    private static final Logger LOG = LoggerFactory.getLogger(BackendConfig.class);
+    public static final String BACKEND_CONFIG_FILE = "analysis.properties";
 
-    protected final Properties config = new Properties();
-    protected final Set<String> missingKeys = new HashSet<>();
+    // INSTANCE FIELDS
 
     // If true, this backend instance should shut itself down after a period of inactivity.
```
[the remark this answers is not in the document]
Comment (abyrd) [reply src/main/java/com/conveyal/analysis/BackendConfig.java:34]: Yes, this is a stray comment - field is now in ClusterConfig.

File: src/main/java/com/conveyal/r5/analyst/cluster/EC2Info.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (abyrd) [reply src/main/java/com/conveyal/r5/analyst/cluster/EC2Info.java:1]: This class is mostly initialized by metadata from the cloud deployment environment, see my PR comments on ClusterUtil.

**Review by ansoncfit — APPROVED** [review]
Looks good

**Pull request merged**